File: app/src/main/java/com/fatbook/fatbookapp/ui/adapters/IngredientAdapter.java
package com.fatbook.fatbookapp.ui.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.fatbook.fatbookapp.R;
import com.fatbook.fatbookapp.core.Ingredient;

import java.util.List;

public class IngredientAdapter extends RecyclerView.Adapter<IngredientAdapter.ViewHolder> {

    private final LayoutInflater inflater;
    private final List<Ingredient> list;

    public IngredientAdapter(Context context, List<Ingredient> list) {
        this.list = list;
        this.inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.rv_ingredient, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Ingredient ingredient = list.get(position);
        holder.tvName.setText(ingredient.getName());

    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        final TextView tvName;
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvName = itemView.findViewById(R.id.editText_dialog_add_ingredient);
        }
    }
}

File: app/src/main/java/online/fatbook/fatbookapp/ui/recipe/adapters/IngredientAdapter.kt
package online.fatbook.fatbookapp.ui.recipe.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.card.MaterialCardView
import online.fatbook.fatbookapp.R
import online.fatbook.fatbookapp.core.recipe.ingredient.Ingredient
import online.fatbook.fatbookapp.databinding.RvIngredientBinding
import online.fatbook.fatbookapp.ui.recipe.listeners.OnIngredientItemClickListener
import online.fatbook.fatbookapp.util.BindableAdapter

class IngredientAdapter(private val context: Context) :
    RecyclerView.Adapter<IngredientAdapter.IngredientItemViewHolder>(),
    BindableAdapter<Ingredient> {

    private var data: List<Ingredient> = ArrayList()
    var listener: OnIngredientItemClickListener? = null
    var selectedIngredient: Ingredient? = Ingredient()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): IngredientItemViewHolder {
        val binding = RvIngredientBinding.inflate(LayoutInflater.from(context), parent, false)
        return IngredientItemViewHolder(binding)
    }

    override fun onBindViewHolder(holder: IngredientItemViewHolder, position: Int) {
        val ingredientItem = data[position]
        holder.bind(ingredientItem)
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun setData(data: List<Ingredient>?) {
        data?.let {
            this.data = it as ArrayList<Ingredient>
            notifyDataSetChanged()
        }
    }

    fun setClickListener(listener: OnIngredientItemClickListener) {
        this.listener = listener
    }

    override fun getItemCount(): Int {
        return data.size
    }

    inner class IngredientItemViewHolder(rvIngredientBinding: RvIngredientBinding) :
        RecyclerView.ViewHolder(rvIngredientBinding.root) {
        private val binding = rvIngredientBinding
        fun bind(ingredientItem: Ingredient?) {
            if (selectedIngredient!!.title == ingredientItem!!.title) {
                selectItem(
                    binding.cardviewRvIngredient,
                    binding.textviewIngredientTitleRvIngredient,
                    binding.textviewIngredientKcalsTitleRvIngredient
                )
            } else {
                unselectItem(
                    binding.cardviewRvIngredient,
                    binding.textviewIngredientTitleRvIngredient,
                    binding.textviewIngredientKcalsTitleRvIngredient
                )
            }

            binding.textviewIngredientTitleRvIngredient.text = ingredientItem.title

            binding.textviewIngredientKcalsTitleRvIngredient.text =
                "${ingredientItem.unitRatio!!.kcal} kcal/${ingredientItem.unitRatio.amount} ${ingredientItem.unitRatio.unit!!.title}"

            if (binding.cardviewRvIngredient.isClickable) {
                binding.cardviewRvIngredient.setOnClickListener {
                    listener!!.onIngredientClick(
                        data.indexOf(selectedIngredient),
                        bindingAdapterPosition,
                        ingredientItem
                    )
                }
            }
        }
    }

    private fun selectItem(
        cardView: MaterialCardView,
        textView: TextView,
        textViewKcalGram: TextView
    ) {
        cardView.isClickable = false
        cardView.setBackgroundResource(R.drawable.select_ingredient_round_corner)
        textView.isSelected = true
        textViewKcalGram.isSelected = true
    }

    private fun unselectItem(
        cardView: MaterialCardView,
        textView: TextView,
        textViewKcalGram: TextView
    ) {
        cardView.isClickable = true
        cardView.setBackgroundResource(R.drawable.unselect_ingredient_round_corner)
        textView.isSelected = false
        textViewKcalGram.isSelected = false
    }
}
